// spdy-transport/lib/spdy-transport/newswapper.js
var swapper = exports;
var crypto = require('crypto');
var signature = require('cookie-signature');
// var MobileDetect = require('mobile-detect'); 
var spdy = require('spdy'); // CookieStoreをspdyに作成しているのでそれにアクセスするため

// 文字に色をつけるよう
var black   = '\u001b[30m';
var red     = '\u001b[31m';
var green   = '\u001b[32m';
var yellow  = '\u001b[33m';
var blue    = '\u001b[34m';
var magenta = '\u001b[35m';
var cyan    = '\u001b[36m';
var white   = '\u001b[37m';

var reset   = '\u001b[0m';
// end

//NOTE: use 'Mobile-Detect'
// var md = new MobileDetect();

//
// ### function uid(len)
// #### @len {Number} UID length
// Generate UID
// NOTE: Copy paste from 'connect'
//
function uid(len) {
	return crypto.randomBytes(Math.ceil(len * 3 / 4)).toString('base64').slice(0, len);
}

//
// ### function parseSetCookie(arrCookie)
// #### @arrCookie {Object} Set-Cookie header value
// Parse Set-Cookie header and return Cookie association array.
// 引数のCookieに対してcookieの値と属性を含むobjectを返す関数
// objCookie[cookieName] = {				
//  value: cookieValue,					
//  attributes: cookieAttributes		
// };
//
function parseSetCookie(arrCookie) {
	var objCookie = {};		//Cookie association array
	for(var i=0; i<arrCookie.length; i++) {
		var cookieAttributes = {}; // object
		var pairs = arrCookie[i].split(';');	
		
		for(var j=0; j<pairs.length; j++) {
			var pair = pairs[j].split('=');	
			
			if(j === 0) {
				var cookieName = pair[0].trim();			
				var cookieValue = (pair[1] || '').trim();	
			} else {
				cookieAttributes[pair[0].trim().toLowerCase()] = (pair[1] || '').trim(); // 属性名：属性値となるobjectを作成
			}
		}
		
		objCookie[cookieName] = {				
			value: cookieValue,					
			attributes: cookieAttributes		
		};
	}
	
	return objCookie;
}

//
// ### function frameSetCookie(objCookie)
// #### @objCookie {Object} Cookie association array
// Build Set-Cookie header based on Cookie association array.
//
function frameSetCookie(objCookie) {
	var arrCookie = new Array();	// Set-Cookie header value

	Object.keys(objCookie).forEach(function(i) {
		var strCookie = (i + '=' + objCookie[i].value + '; ');

		Object.keys(objCookie[i].attributes).forEach(function(j) {
			if(j === 'secure' || j === 'httponly') {
				strCookie += (j + '; ');
			} else {
				strCookie += (j + '=' + objCookie[i].attributes[j] + '; ');
			}
		});

		arrCookie.push(strCookie.slice(0, -2));
	});
	
	return arrCookie;
}

//
// ### function parseCookie(strCookie)
// #### @strCookie {String} Cookie header value
// Parse Cookie header and return Cookie association array
//
function parseCookie(strCookie) {
	var objCookie = {};		// Cookie association array
	var pairs = strCookie.split(';');	
	
	for(var i=0; i<pairs.length; i++) {
		var pair = pairs[i].split('=');	
		objCookie[pair[0].trim()] = (pair[1] || '').trim();
	}
	
	return objCookie;
}

//
// ### function frameCookie(objCookie)
// #### @objCookie {Object} Cookie association array
// Build Cookie header based on Cookie association array.
//
function frameCookie(objCookie) {
	var strCookie = '';	// Cookie header value

	Object.keys(objCookie).forEach(function(i) {
		strCookie += (i + '=' + objCookie[i] + '; ');
	});

	return strCookie.slice(0, -2);	
}

//
// ### function Swapper()
// Swapper @constructor
// Swapperのコンストラクタ？
//
function Swapper() {
	// Random character strings used to generate session ID.
  this.secret = uid(12);
  this.swappertime = 0;
	this.SSCookies = {};
	this.CSCookies = {};
}

//
// ### function create()
// @constructor wrapper
// Swapperインスタンスを生成？
//
exports.create = function create() {
	return new Swapper();
};

// add in ES3M
//
// ### function storefeatures
// ユーザの特徴をstoreに保存する。
//
function storefeatures(headers, cookies, j, ip, self){

  let flag = 0; // refererが保存されているかどうかの確認用
  //console.log('flag:' + flag);

  //console.log('あああああ')
  // refererが存在しない　すなわち　最初の通信
  if (!spdy.store[cookies[j]].hasOwnProperty('referer')) {
    //console.log('spdy.store[cookies[j]].referer:' + spdy.store[cookies[j]].referer);
    spdy.store[cookies[j]].referer = {};
    //console.log('spdy.store[cookies[j]].referer:' + spdy.store[cookies[j]].referer);
    spdy.store[cookies[j]].referer.re1 = {};
    spdy.store[cookies[j]].referer.re1.br = headers['referer'];
    spdy.store[cookies[j]].referer.re1.nr = [headers[":scheme"] + "://" + headers[":authority"] + headers[":path"]];
    flag = 1;
    console.log('最初の通信のrefererを保存')
    /*
    console.log('flag:' + flag);
    console.log('spdy.store[cookies[j]].referer:' + spdy.store[cookies[j]].referer);
    console.log('spdy.store[cookies[j]].referer.re1:' + spdy.store[cookies[j]].referer.re1);
    console.log('spdy.store[cookies[j]].referer.re1.br:' + spdy.store[cookies[j]].referer.re1.br);
    console.log('spdy.store[cookies[j]].referer.re1.nr:' + spdy.store[cookies[j]].referer.re1.nr);
    console.log('いいいいい')
    console.log(spdy.store[cookies[j]]);
    */
  }
  //console.log('ううううう')

  // refererが登録されているときに実行（初回以外）
  if(flag === 0){
    // 通常時のrefererの処理
    console.log('flagが0の時の処理');
    Object.keys(spdy.store[cookies[j]].referer).forEach(function(k){
      for(i=0; i<spdy.store[cookies[j]].referer[k].nr.length; i++){
        if(headers['referer'] === spdy.store[cookies[j]].referer[k].nr[i]){
          /* Object.keysで呼び出したプロパティはドット記法(.)ではなく、ブラケット記法([])を使わなきゃいけないらしい
          console.log('k:' + k);
          console.log('spdy.store[cookies[j]].referer.re1.nr:' + spdy.store[cookies[j]].referer.re1.nr);
          console.log('spdy.store[cookies[j]].referer[k]:' + spdy.store[cookies[j]].referer[k]);
          console.log('spdy.store[cookies[j]].referer.k:' + spdy.store[cookies[j]].referer.k);
          console.log('spdy.store[cookies[j]].referer[k].nr:' + spdy.store[cookies[j]].referer[k].nr);
          console.log('spdy.store[cookies[j]].referer.k.nr:' + spdy.store[cookies[j]].referer.k.nr);
          */
          // 既に処理した場合、他の物は処理しない
          if(flag === 2){
            return;
          }
          
          console.log('通常時のrefererの処理')
          flag = 2;
          spdy.store[cookies[j]].referer[k].br = spdy.store[cookies[j]].referer[k].nr[i];
          spdy.store[cookies[j]].referer[k].nr = [headers[":scheme"] + "://" + headers[":authority"] + headers[":path"]];
        }
      }
    })
    // リダイレクト時のrefererの処理
    // brのrefererが一致するときはプロパティを追加する処理が必要
    if(flag === 0){
      // brのrefererが一致する場合の時の処理
      Object.keys(spdy.store[cookies[j]].referer).forEach(function(k){
        Object.keys(spdy.store[cookies[j]].referer).forEach(function(l){
          if(spdy.store[cookies[j]].referer[k].br === spdy.store[cookies[j]].referer[l].br){
            // nrの値を一方に追加
            spdy.store[cookies[j]].referer[k].nr = spdy.store[cookies[j]].referer[k].nr.concat(spdy.store[cookies[j]].referer[l].nr);
            // 重複を削除
            spdy.store[cookies[j]].referer[k].nr = spdy.store[cookies[j]].referer[k].nr.filter((x, i, self) => self.indexOf(x) === i);
            // 追加したプロパティをもう片方にコピー
            spdy.store[cookies[j]].referer[l] = JSON.parse(JSON.stringify(spdy.store[cookies[j]].referer[k]));
          }
        })
      })
      // リダイレクト時のrefererの処理
      Object.keys(spdy.store[cookies[j]].referer).forEach(function(k){
        if (headers['referer'] === spdy.store[cookies[j]].referer[k].br){

          // 既に処理した場合、他の物は処理しない
          if(flag === 2){
            return;
          }

          console.log('リダイレクト時のrefererの処理')
          spdy.store[cookies[j]].referer[k].nr = [headers[":scheme"] + "://" + headers[":authority"] + headers[":path"]];
          flag = 2
        }
      })
    }
  }

  // 一致するrefererが存在しない かつ　swappertimeが一致する
  if(flag === 0 && spdy.store[cookies[j]].swappertime === self.swappertime){
    let temp = 're' + (Object.keys(spdy.store[cookies[j]].referer).length + 1);

    console.log("追加のreferer（temp）:" + temp);

    spdy.store[cookies[j]].referer[temp] = {};
    spdy.store[cookies[j]].referer[temp].br = headers['referer'];
    spdy.store[cookies[j]].referer[temp].nr = [headers[":scheme"] + "://" + headers[":authority"] + headers[":path"]];

  }

  console.log('storefeaturesでのrefereの記録');
  console.log('haeders[referer]:' + headers['referer']);
  console.table(spdy.store[cookies[j]].referer);
       
  //対応する送信元IPアドレスが記憶されていない場合、送信元IPアドレスを記憶
  if(!spdy.store[cookies[j]].hasOwnProperty('ip')){
    spdy.store[cookies[j]].ip = ip;
  }

  //対応するUser-Agentが記憶されていない場合、User-Agentを記憶
  if(!spdy.store[cookies[j]].hasOwnProperty('ua')){
    spdy.store[cookies[j]].ua = headers["user-agent"];
  }

  //対応するアクセス時間が記憶されていない場合、アクセス時間を記憶
  if(!spdy.store[cookies[j]].hasOwnProperty('time')){
    spdy.store[cookies[j]].time = Date.now();
  }
}
// end

// add in ES3M
//
// ### function referercheck
// refererが一致するか確認する
//
function referercheck(headers, k){

  console.log('referercheckを実行');

  let flag = 0;

  /*
  Object.keys(spdy.store[k].referer).forEach(function(l){
    if (headers['referer'] === spdy.store[k].referer[l].nr){
      console.log('referer.nrが一致');
      flag = 1;
    }
  })

  if(flag === 0){
    Object.keys(spdy.store[k].referer).forEach(function(l){
      if (headers['referer'] === spdy.store[k].referer[l].br){
        console.log('referer.brが一致');
        flag = 1;
      }
    })
  }
  */

  // storefeaturesでrefererの確認を行い、refererが一致したものをbrに入れているので
  // その確認を行えばよい？

  Object.keys(spdy.store[k].referer).forEach(function(l){
   if (headers['referer'] === spdy.store[k].referer[l].br){
     console.log('referer.brが一致');
     flag = 1;
    }
  })

  if(flag === 1){
    return true;
  }

  console.log("一致するrefererは存在しない")
  return false;

}

//
// ### function SwapSforC(headers)
// #### @headers {Object} Response header
// Swap SS Cookie for CS Cookie
// SSCookieをCSCookieに変換するメソッドswapSforCを定義。
// functionの後ろのswapSforC必要？
//
Swapper.prototype.swapSforC = function swapSforC(headers, statusCode) {

  console.log(blue + "swapSforC関数の実行" + reset);
  //console.log("statusCode:" + statusCode);

  var self = this;
  for(var i in headers) {

    var headerName = (i || ''); // headerNameにi(ヘッダ名)か''(空文字列)を代入
    if(headerName.toLowerCase() !== 'set-cookie') continue; // headerNameがset-cookieならば下へ続く
    console.log("set-cookieヘッダを確認")

    var cookies = parseSetCookie(headers[i]); // set-cookieをオブジェクト化する
    //console.log("parseSetCookie(serverから送られてきたcookie):" + cookies); // cookiesの中身の確認

    // Object.keys(cookies)はcookiesのnamesプロパティの配列を返す
    // [value, attributes]
    // Array.forEach(function(j){});は与えられた関数を配列の各要素に対して1度ずつ実行する
    Object.keys(cookies).forEach(function(j) { 
      
      //Cookie flag
      //Cookieの変換を行うかの基準
      //Cookieの属性にexpiresがないかつmax-ageがないかつhttponlyがあるときにtrueを保持
      var criteria = (
        !cookies[j].attributes['expires'] &&
        !cookies[j].attributes['max-age'] &&
        cookies[j].attributes.hasOwnProperty('httponly')
      );

      //console.log("cookieの属性:" + cookies[j].attributes);//0807追加
      
      //console.log("provisonalreferer:" + spdy.store[cookies[j]].provisonalreferer); provisonalrefererの手前がundefinedらしい
      //console.log(cookies[j].attributes['path']);//0807追加 パスは取れる　これをrefererにいれれればよさそう
      //console.log(spdy.store[cookies[j]].ua);//0807追加　spdy.storeにrefererを入れたいがなぜかcan't set headers after they are sentエラーが起こる
      //refererを記憶
      //spdy.store[cookies[j]].referer = cookies[j].attributes['path'];//0807追加
      //console.log("referer:" + spdy.store[cookies[j]].referer);//0807追加

      //statusCodeが200ならばguessrefererにprovisonalrefererを代入
      //if(statusCode === 200) {
      //  spdy.store[self.CSCookies[j]].guessreferer = spdy.store[self.CSCookies[j]].provisonalreferer;
      //}

      //console.table(spdy.store[self.CSCookies[j]]);
      console.log(JSON.stringify(spdy.store[self.CSCookies[j]], null, "  "));
      console.log(JSON.stringify(spdy.store, null, "  "));

      if(self.SSCookies.hasOwnProperty(j) && criteria) {
        // The SS Cookie is pre-stored,　
        // Cookie store flag is true and 
        // it is not an access from a mobile device.
        console.log("SSCookieが保存されており、flagが立っている場合に実行し、CSCookieを読み込む");

        // Load CS Cookie
        cookies[j].value = self.CSCookies[j];

      }else if(self.SSCookies.hasOwnProperty(j)){
        // The SS Cookie is pre-stored and Cookie store flag is false.
        // Delete CS Cookie
        console.log("SSCookieが保存されており、flagが立っていない場合に実行し、CSCookieを削除する");

        delete self.CSCookies[j]; 

      }else if(criteria){
        // The SS Cookie isn't stored,
        // Cookie store flag is true and 
        // it is not an access from a mobile device.
        console.log("SSCookieが保存されておらず、flagが立っている場合に実行し、CSCookieを作成しSSCookieとともに保存する");

        var ss_value = cookies[j].value;
        self.SSCookies[j] = ss_value;

        var cs_value = 'c:' + signature.sign(uid(24), self.secret);

        self.CSCookies[j] = cs_value;

        spdy.store[cs_value] = {"ss" : ss_value};

        cookies[j].value = cs_value;

        // TCPコネクションが同じであるかどうかの確認用
        self.swappertime = Date.now();
        spdy.store[cs_value].swappertime = self.swappertime;
      }
    });

    // Rebuild Set-Cookie header
    headers[i] = frameSetCookie(cookies);
    break;
  }
};

//
// ### function SwapCforS(headers)
// #### @headers {Object} request header
// Swap CS Cookie for SS Cookie
// CSCookieをSSCookieに変換するメソッドswapSforCを定義。
// functionの後ろのswapCforS必要？
//
Swapper.prototype.swapCforS = function swapCforS(headers, ip) {

  console.log(red + "swapCforS関数の実行" + reset);

  var self = this;

  //ヘッダからCookieヘッダを検索
  for(var i in headers) {

    /*
    // 宛先に"."がある時は実行しない（画像等のリクエストのため）
    if(headers[':path'].indexOf('.') !== -1) {
      console.log("headers[':path']:" + headers[':path']);
      console.log('宛先に"."があるため、画像等のリクエストと判断し');

      Object.keys(cookies).forEach(function(j){
        cookies[j] = self.SSCookies[j]; 
      })
      headers[i] = frameCookie(cookies);
      return;
    }
    */
    
    //console.log("ヘッダからcookieヘッダを検索");

    var headerName = (i || '');

    //Cookieヘッダ以外は終了
    if(headerName.toLowerCase() !== 'cookie') {
      //console.log("ヘッダ名:" + headerName.toLowerCase());
      //console.log("cookieヘッダじゃないので次のヘッダへ");
      continue; 
    }

    console.log(red + "宛先は" + headers[":scheme"] + "://" + headers[":authority"] + headers[":path"] + reset);

    //Cookieヘッダに含まれるCookieの連想配列を生成
    var cookies = parseCookie(headers[i]);
    //console.log("CS Cookie:" + headers[i]);

    if(headers[':path'].indexOf('.') !== -1) {
      console.log("headers[':path']:" + headers[':path']);
      console.log('宛先に"."があるため、画像等のリクエストと判断し');

      Object.keys(cookies).forEach(function(j){
        cookies[j] = self.SSCookies[j]; 
      })
      headers[i] = frameCookie(cookies);
      return;
    }

    // jは通信に含まれている各cookieを示す？
    Object.keys(cookies).forEach(function(j) {

      // 該当しないCookieを持っていない場合これより下は意味ない？
      // CSCookieを初めて利用するときユーザ情報が記憶されていないので記憶する。
      // また、refererの更新を行う。
      // 実行する条件はStoreが該当するCSCookieを持っていた場合。
      // 一度記憶すれば、IPアドレスとUAは更新しない。
      if(spdy.store.hasOwnProperty(cookies[j])){
        console.log("storefeaturesを実行");
        storefeatures(headers, cookies, j, ip, self);
      }

      // TCPコネクションが切れているかどうかをSwapperで確認
      // CS CookieがSwapperに記憶されている場合
      if(self.CSCookies.hasOwnProperty(j)){

        console.log("CScookieがSwapperに記憶されている");
        
        // 記憶されているCSCookieとCookieヘッダに含まれるCSCookieが一致する場合
        if(cookies[j] === self.CSCookies[j]){

          // Storeに記憶した情報を更新
          spdy.store[cookies[j]].time = Date.now();

          // 対応するSSCookieと交換
          cookies[j] = self.SSCookies[j]; 
          console.log("SSCookieと交換");

        // 記憶されているCS CookieとCookieヘッダに含まれるCS Cookieが一致しない場合
        // 一致しないとかある？本来あるはずのないCookieがあるということ？
        }else{

          // Cookieを削除
          delete cookies[j];
          console.log("CSCookie?を削除");

          console.log("CSCookieがSwapperに記憶されているが、一致しないので削除")
        }

      //CS CookieがSwapperに記憶されていない場合
      }else{

        console.log("CScookieがSwapperに記憶されていない");//0805追加

        // TCPコネクションが切断されたのでswappertimeの更新
        self.swappertime = Date.now();
        spdy.store[cookies[j]].swappertime = self.swappertime;


        // swappertimeの更新
        console.log("swappertimeの更新")
        self.swappertime = Date.now();
        spdy.store[cookies[j]].swappertime = self.swappertime;
        
        // Storeに記憶されているCS Cookieを検索
        Object.keys(spdy.store).forEach(function(k) {
	     
          // 対象のCS Cookieが記憶されていない場合、終了
          if(cookies[j] !== k) {
            return;
          }

          // 対象のCS Cookieが記憶されている場合
          if(cookies[j] === k) {

            // 前回のアクセスから10分以内、
            // かつ送信元IPアドレスがStoreに記憶した送信元IPアドレスと一致する
            // かつUser-AgentがStoreに記憶したUser-Agentと一致する場合
	          // かつrefererがStoreに記憶したrefererと一致する場合
            if(Date.now() < spdy.store[k].time + 10*60*1000 
              && ip === spdy.store[k].ip
              && headers['user-agent'] === spdy.store[k].ua
	            && referercheck(headers, k)){//0807追記
 
              //SwapperにCS CookieとSS Cookieを記憶
              self.CSCookies[j] = cookies[j];
              self.SSCookies[j] = spdy.store[k].ss;

              //対応するSS Cookieと交換
              cookies[j] = spdy.store[k].ss;
              console.log("対応するSSCookieと交換")

              //Storeに記憶した情報を更新
              spdy.store[k].time = Date.now();

            //条件を1つでも満たさない場合
            }else{

              //Storeに記憶してある対象のCookieを削除
              delete k;
              console.log("CSCookie？を削除")
            }
          }
        });
      }

      //Cookieヘッダを再構築
      headers[i] = frameCookie(cookies);
      //console.log("SS Cookie:" + headers[i]);

    });
  }
};
